xbmc/input/touch/generic/GenericTouchActionHandler.cpp

```diff
@@ -179,7 +179,7 @@ void CGenericTouchActionHandler::sendEvent(int actionId, float x, float y, float
   newEvent.touch.y3 = y3;
   newEvent.touch.pointers = pointers;
 
-  CWinEvents::MessagePush(&newEvent);
+  g_application.OnEvent(newEvent);
```

Comment:
Is it OK to call CApplication::OnEvent from any thread?

Reply:
it has to be ok. Ideally OnEvent should not block the caller, in particular it should not hijack the render thread for synchronous processing of the event. something for one of the next iterations

Reply:
Then touch on Wayland might break in this iteration, because this will absolutely be called from the Wayland event handling thread. If you are aware of it then it's fine.

Reply:
if an event required the main thread to be processed, it was already broken in terms of design.

Reply:
Might be, but we still need to do something about it now that `OnEvent` is not only called from the main thread. I'm not sure what the solution is. Do we have to make sure that every code path through `OnEvent` is thread-safe?

Reply:
the current design is as follows: if something requires to be run on the main thread, it has to use ApplicationMessanger. That has been the original purpose of that class.
now we have to distinguish between the main thread and the render thread. ApplicationMessenger is for messages in direction of the main tread. Windowing will encapsulate the renderthread and the interface will hide the details. That means it will take care what calls it will pass to the render thread internally.

Reply:
btw: there is another design flaw associated with CGenericTouchInputHandler: it is one of those ugly global singletons. Kill the global and embed an instance into Wayland. Then you can handle those events internally.